File: src/sc_neurocore/accel/go/services/neurogrid.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go candidate-first RK4 service for neurogrid

package services

import "math"

// NeuroGridNeuronState holds the reduced Neurogrid two-compartment state.
type NeuroGridNeuronState struct {
	VS         float64
	VD         float64
	TauS       float64
	TauD       float64
	GC         float64
	DeltaT     float64
	VRest      float64
	VThreshold float64
	VPeak      float64
	VReset     float64
	Dt         float64
}

// NewNeuroGridNeuron creates a new NeuroGridNeuron neuron with default parameters.
func NewNeuroGridNeuron() *NeuroGridNeuronState {
	return &NeuroGridNeuronState{
		VS:         -65.0,
		VD:         -65.0,
		TauS:       20.0,
		TauD:       50.0,
		GC:         0.5,
		DeltaT:     2.0,
		VRest:      -65.0,
		VThreshold: -50.0,
		VPeak:      20.0,
		VReset:     -65.0,
		Dt:         0.1,
	}
}

func neurogridFinite(values ...float64) bool {
	for _, value := range values {
		if math.IsNaN(value) || math.IsInf(value, 0) {
			return false
		}
	}
	return true
}

func (s *NeuroGridNeuronState) valid() bool {
	return neurogridFinite(s.VS, s.VD, s.TauS, s.TauD, s.GC, s.DeltaT, s.VRest, s.VThreshold, s.VPeak, s.VReset, s.Dt) &&
		s.TauS > 0.0 && s.TauD > 0.0 && s.DeltaT > 0.0 && s.Dt > 0.0 && s.GC >= 0.0
}

func (s *NeuroGridNeuronState) derivatives(vs float64, vd float64, current float64) (float64, float64) {
	vsEff := math.Min(vs, s.VPeak)
	dVD := (-(vd - s.VRest) + current - s.GC*(vd-vsEff)) / s.TauD
	expArg := math.Min((vsEff-s.VThreshold)/s.DeltaT, 20.0)
	expTerm := s.DeltaT * math.Exp(expArg)
	dVS := (-(vsEff - s.VRest) + expTerm + s.GC*(vd-vsEff)) / s.TauS
	return dVS, dVD
}

func (s *NeuroGridNeuronState) rk4Substep(vs float64, vd float64, current float64) (float64, float64) {
	dt := s.Dt
	k1vs, k1vd := s.derivatives(vs, vd, current)
	k2vs, k2vd := s.derivatives(vs+0.5*dt*k1vs, vd+0.5*dt*k1vd, current)
	k3vs, k3vd := s.derivatives(vs+0.5*dt*k2vs, vd+0.5*dt*k2vd, current)
	k4vs, k4vd := s.derivatives(vs+dt*k3vs, vd+dt*k3vd, current)
	nextVS := vs + dt*(k1vs+2.0*k2vs+2.0*k3vs+k4vs)/6.0
	nextVD := vd + dt*(k1vd+2.0*k2vd+2.0*k3vd+k4vd)/6.0
	return nextVS, nextVD
}

// Step advances the neuron by one candidate-first RK4 timestep.
func (s *NeuroGridNeuronState) Step(iExt float64) int {
	if !neurogridFinite(iExt) || !s.valid() {
		return 0
	}
	nextVS, nextVD := s.rk4Substep(s.VS, s.VD, iExt)
	if !neurogridFinite(nextVS, nextVD) {
		return 0
	}
	s.VD = nextVD
	if nextVS >= s.VPeak {
		s.VS = s.VReset
		return 1
	}
	s.VS = nextVS
	return 0
}

// SimulateNeuroGridNeuron runs the neuron for n steps.
func SimulateNeuroGridNeuron(nSteps int, iExt float64) ([]float64, int) {
	s := NewNeuroGridNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for t := 0; t < nSteps; t++ {
		result := s.Step(iExt)
		trace[t] = s.VS
		if result > 0 {
			spikes++
		}
	}
	return trace, spikes
}

File: src/sc_neurocore/accel/julia/neurons/neurogrid.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia candidate-first RK4 for neurogrid

module NeurogridAccel

export step!, simulate, NeuroGridNeuronState

mutable struct NeuroGridNeuronState
    v_s::Float64
    v_d::Float64
    tau_s::Float64
    tau_d::Float64
    g_c::Float64
    delta_t::Float64
    v_rest::Float64
    v_threshold::Float64
    v_peak::Float64
    v_reset::Float64
    dt::Float64
end

function NeuroGridNeuronState()
    NeuroGridNeuronState(-65.0, -65.0, 20.0, 50.0, 0.5, 2.0, -65.0, -50.0, 20.0, -65.0, 0.1)
end

@inline function _finite(values...)
    return all(isfinite, values)
end

@inline function _valid(s::NeuroGridNeuronState)
    return _finite(s.v_s, s.v_d, s.tau_s, s.tau_d, s.g_c, s.delta_t, s.v_rest, s.v_threshold, s.v_peak, s.v_reset, s.dt) &&
        s.tau_s > 0.0 && s.tau_d > 0.0 && s.delta_t > 0.0 && s.dt > 0.0 && s.g_c >= 0.0
end

@inline function _derivatives(s::NeuroGridNeuronState, v_s::Float64, v_d::Float64, current::Float64)
    v_s_eff = min(v_s, s.v_peak)
    dv_d = (-(v_d - s.v_rest) + current - s.g_c * (v_d - v_s_eff)) / s.tau_d
    exp_arg = min((v_s_eff - s.v_threshold) / s.delta_t, 20.0)
    exp_term = s.delta_t * exp(exp_arg)
    dv_s = (-(v_s_eff - s.v_rest) + exp_term + s.g_c * (v_d - v_s_eff)) / s.tau_s
    return dv_s, dv_d
end

@inline function _rk4_substep(s::NeuroGridNeuronState, v_s::Float64, v_d::Float64, current::Float64)
    dt = s.dt
    k1s, k1d = _derivatives(s, v_s, v_d, current)
    k2s, k2d = _derivatives(s, v_s + 0.5 * dt * k1s, v_d + 0.5 * dt * k1d, current)
    k3s, k3d = _derivatives(s, v_s + 0.5 * dt * k2s, v_d + 0.5 * dt * k2d, current)
    k4s, k4d = _derivatives(s, v_s + dt * k3s, v_d + dt * k3d, current)
    return (
        v_s + dt * (k1s + 2.0 * k2s + 2.0 * k3s + k4s) / 6.0,
        v_d + dt * (k1d + 2.0 * k2d + 2.0 * k3d + k4d) / 6.0,
    )
end

function step!(s::NeuroGridNeuronState, I_ext::Float64=0.0)
    if !_finite(I_ext) || !_valid(s)
        return 0
    end
    next_v_s, next_v_d = _rk4_substep(s, s.v_s, s.v_d, I_ext)
    if !_finite(next_v_s, next_v_d)
        return 0
    end
    s.v_d = next_v_d
    if next_v_s >= s.v_peak
        s.v_s = s.v_reset
        return 1
    end
    s.v_s = next_v_s
    return 0
end

function simulate(n_steps::Int=1000; I_ext::Float64=100.0)
    s = NeuroGridNeuronState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext)
        trace[t] = s.v_s
        if result isa Number && result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

end # module NeurogridAccel
